src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -295,120 +297,149 @@ public static string Combine(params string[] paths)
-                sb.Append(path1)
-                  .Append(PathInternal.DirectorySeparatorChar)
-                  .Append(path2)
-                  .Append(PathInternal.DirectorySeparatorChar)
-                  .Append(path3);
-                return StringBuilderCache.GetStringAndRelease(sb);
-            }
+            if (first.Length == 0)
+                return CombineNoChecks(second, third);
+            if (second.Length == 0)
+                return CombineNoChecks(first, third);
+            if (third.Length == 0)
+                return CombineNoChecks(first, second);
+
+            if (IsPathRooted(third))
+                return new string(third);
+            if (IsPathRooted(second))
+                return CombineNoChecks(second, third);
+
+            return CombineNoChecksInternal(first, second, third);            
         }
 
-        private static string CombineNoChecks(string path1, string path2, string path3, string path4)
+        private static string CombineNoChecks(ReadOnlySpan<char> first, ReadOnlySpan<char> second, ReadOnlySpan<char> third, ReadOnlySpan<char> fourth)
         {
-            if (path1.Length == 0)
-                return CombineNoChecks(path2, path3, path4);
-            if (path2.Length == 0)
-                return CombineNoChecks(path1, path3, path4);
-            if (path3.Length == 0)
-                return CombineNoChecks(path1, path2, path4);
-            if (path4.Length == 0)
-                return CombineNoChecks(path1, path2, path3);
-
-            if (IsPathRooted(path4))
-                return path4;
-            if (IsPathRooted(path3))
-                return CombineNoChecks(path3, path4);
-            if (IsPathRooted(path2))
-                return CombineNoChecks(path2, path3, path4);
-
-            bool hasSep1 = PathInternal.IsDirectoryOrVolumeSeparator(path1[path1.Length - 1]);
-            bool hasSep2 = PathInternal.IsDirectoryOrVolumeSeparator(path2[path2.Length - 1]);
-            bool hasSep3 = PathInternal.IsDirectoryOrVolumeSeparator(path3[path3.Length - 1]);
-
-            if (hasSep1 && hasSep2 && hasSep3)
+            if (first.Length == 0)
+                return CombineNoChecks(second, third, fourth);
+            if (second.Length == 0)
+                return CombineNoChecks(first, third, fourth);
+            if (third.Length == 0)
+                return CombineNoChecks(first, second, fourth);
+            if (fourth.Length == 0)
+                return CombineNoChecks(first, second, third);
+
+            if (IsPathRooted(fourth))
+                return new string(fourth);
+            if (IsPathRooted(third))
+                return CombineNoChecks(third, fourth);
+            if (IsPathRooted(second))
+                return CombineNoChecks(second, third, fourth);
+
+            return CombineNoChecksInternal(first, second, third, fourth);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I'm not really sure how the JIT will handle the inlining here. I'm a little worried that we end up with a ton of copies. I don't think they're necessary, I'd remove them.